File: backend/handlers/shares.go
package handlers

import (
	"context"
	"time"

	"github.com/kutup/backend/middleware"
	"github.com/kutup/backend/services"
	"github.com/kutup/backend/utils"
	"github.com/gofiber/fiber/v2"
	"github.com/jackc/pgx/v5/pgxpool"
)

type SharesHandler struct {
	DB      *pgxpool.Pool
	Storage *services.StorageService
}

// CreatePublicShare creates a public share link.
// CRITICAL: The linkKey is NEVER sent here — it lives only in the URL #fragment.
// We only store the collection key already encrypted with that linkKey by the client.
// @Summary      Create a public share link
// @Tags         Public Shares
// @Accept       json
// @Produce      json
// @Security     BearerAuth
// @Param        body  body      CreateShareRequest  true  "Share details"
// @Success      201   {object}  CreateShareResult
// @Failure      400   {object}  ErrorResponse
// @Failure      401   {object}  ErrorResponse
// @Failure      403   {object}  ErrorResponse
// @Router       /share [post]
func (h *SharesHandler) CreatePublicShare(c *fiber.Ctx) error {
	userID := middleware.UserID(c)

	var req struct {
		ShareType                   string  `json:"shareType"` // "collection" or "file"
		TargetID                    string  `json:"targetId"`
		EncryptedCollectionKey      string  `json:"encryptedCollectionKey"`
		EncryptedCollectionKeyNonce string  `json:"encryptedCollectionKeyNonce"`
		ExpiresInHours              *int    `json:"expiresInHours"`
	}
	if err := c.BodyParser(&req); err != nil {
		return c.Status(400).JSON(fiber.Map{"error": "invalid request"})
	}

	// Verify user owns the target
	if !h.userOwnsTarget(c.Context(), userID, req.ShareType, req.TargetID) {
		return c.Status(403).JSON(fiber.Map{"error": "forbidden"})
	}

	token, err := utils.RandomToken(32)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "internal error"})
	}

	var expiresAt *time.Time
	if req.ExpiresInHours != nil {
		t := time.Now().Add(time.Duration(*req.ExpiresInHours) * time.Hour)
		expiresAt = &t
	}

	var id string
	err = h.DB.QueryRow(context.Background(), `
		INSERT INTO public_shares (share_type, target_id, token,
		                           encrypted_collection_key, encrypted_collection_key_nonce,
		                           expires_at)
		VALUES ($1,$2,$3,$4,$5,$6)
		RETURNING id
	`, req.ShareType, req.TargetID, token,
		req.EncryptedCollectionKey, req.EncryptedCollectionKeyNonce,
		expiresAt,
	).Scan(&id)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "internal error"})
	}

	return c.Status(201).JSON(fiber.Map{
		"id":    id,
		"token": token,
	})
}

// GetPublicShare returns encrypted key material for a public share.
// No auth required — anyone with the token can get the ciphertext.
// Without the linkKey (URL fragment), the ciphertext is useless.
// @Summary      Get public share metadata
// @Tags         Public Shares
// @Produce      json
// @Param        token  path      string  true  "Share token"
// @Success      200    {object}  PublicShareResponse
// @Failure      404    {object}  ErrorResponse
// @Failure      410    {object}  ErrorResponse  "Link expired"
// @Router       /share/{token} [get]
func (h *SharesHandler) GetPublicShare(c *fiber.Ctx) error {
	token := c.Params("token")

	var share struct {
		ID                          string     `json:"id"`
		ShareType                   string     `json:"shareType"`
		TargetID                    string     `json:"targetId"`
		EncryptedCollectionKey      *string    `json:"encryptedCollectionKey"`
		EncryptedCollectionKeyNonce *string    `json:"encryptedCollectionKeyNonce"`
		ExpiresAt                   *time.Time `json:"expiresAt"`
	}

	err := h.DB.QueryRow(context.Background(), `
		SELECT id, share_type, target_id,
		       encrypted_collection_key, encrypted_collection_key_nonce,
		       expires_at
		FROM public_shares
		WHERE token = $1
	`, token).Scan(
		&share.ID, &share.ShareType, &share.TargetID,
		&share.EncryptedCollectionKey, &share.EncryptedCollectionKeyNonce,
		&share.ExpiresAt,
	)
	if err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "not found"})
	}

	if share.ExpiresAt != nil && time.Now().After(*share.ExpiresAt) {
		return c.Status(410).JSON(fiber.Map{"error": "link expired"})
	}

	return c.JSON(share)
}

// ListPublicShareFiles lists files in a public share (no auth, no decryption).
// @Summary      List files in a public share
// @Tags         Public Shares
// @Produce      json
// @Param        token  path      string  true  "Share token"
// @Success      200    {array}   FileRow
// @Failure      400    {object}  ErrorResponse
// @Failure      404    {object}  ErrorResponse
// @Failure      410    {object}  ErrorResponse  "Link expired"
// @Router       /share/{token}/files [get]
func (h *SharesHandler) ListPublicShareFiles(c *fiber.Ctx) error {
	token := c.Params("token")

	var targetID, shareType string
	var expiresAt *time.Time
	err := h.DB.QueryRow(context.Background(),
		`SELECT target_id, share_type, expires_at FROM public_shares WHERE token = $1`,
		token,
	).Scan(&targetID, &shareType, &expiresAt)
	if err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "not found"})
	}

	if expiresAt != nil && time.Now().After(*expiresAt) {
		return c.Status(410).JSON(fiber.Map{"error": "link expired"})
	}

	if shareType != "collection" {
		return c.Status(400).JSON(fiber.Map{"error": "not a collection share"})
	}

	rows, err := h.DB.Query(context.Background(), `
		SELECT id, collection_id, encrypted_metadata, metadata_nonce,
		       encrypted_file_key, file_key_nonce, encrypted_size_bytes, created_at
		FROM files WHERE collection_id = $1
		ORDER BY created_at DESC
	`, targetID)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "internal error"})
	}
	defer rows.Close()

	type FileRow struct {
		ID                 string `json:"id"`
		CollectionID       string `json:"collectionId"`
		EncryptedMetadata  string `json:"encryptedMetadata"`
		MetadataNonce      string `json:"metadataNonce"`
		EncryptedFileKey   string `json:"encryptedFileKey"`
		FileKeyNonce       string `json:"fileKeyNonce"`
		EncryptedSizeBytes int64  `json:"encryptedSizeBytes"`
		CreatedAt          string `json:"createdAt"`
	}

	var files []FileRow
	for rows.Next() {
		var f FileRow
		if err := rows.Scan(
			&f.ID, &f.CollectionID, &f.EncryptedMetadata, &f.MetadataNonce,
			&f.EncryptedFileKey, &f.FileKeyNonce, &f.EncryptedSizeBytes, &f.CreatedAt,
		); err != nil {
			continue
		}
		files = append(files, f)
	}
	if files == nil {
		files = []FileRow{}
	}
	return c.JSON(files)
}

// DownloadPublicShareFile returns a presigned URL for a public share file.
// @Summary      Get presigned download URL for a public share file
// @Tags         Public Shares
// @Produce      json
// @Param        token   path      string  true  "Share token"
// @Param        fileId  path      string  true  "File UUID"
// @Success      200     {object}  DownloadURLResponse
// @Failure      403     {object}  ErrorResponse
// @Failure      404     {object}  ErrorResponse
// @Failure      410     {object}  ErrorResponse  "Link expired"
// @Router       /share/{token}/download/{fileId} [get]
func (h *SharesHandler) DownloadPublicShareFile(c *fiber.Ctx) error {
	token := c.Params("token")
	fileID := c.Params("fileId")

	// Validate share is still valid
	var targetID, shareType string
	var expiresAt *time.Time
	err := h.DB.QueryRow(context.Background(),
		`SELECT target_id, share_type, expires_at FROM public_shares WHERE token = $1`,
		token,
	).Scan(&targetID, &shareType, &expiresAt)
	if err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "not found"})
	}

	if expiresAt != nil && time.Now().After(*expiresAt) {
		return c.Status(410).JSON(fiber.Map{"error": "link expired"})
	}

	// Verify file belongs to the shared collection
	var storagePath string
	var collID string
	err = h.DB.QueryRow(context.Background(),
		`SELECT storage_path, collection_id FROM files WHERE id = $1`,
		fileID,
	).Scan(&storagePath, &collID)
	if err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "not found"})
	}

	if shareType == "collection" && collID != targetID {
		return c.Status(403).JSON(fiber.Map{"error": "forbidden"})
	}
	if shareType == "file" && fileID != targetID {
		return c.Status(403).JSON(fiber.Map{"error": "forbidden"})
	}

	url, err := h.Storage.PresignedDownload(c.Context(), storagePath)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "internal error"})
	}

	return c.JSON(fiber.Map{"url": url})
}

func (h *SharesHandler) userOwnsTarget(ctx context.Context, userID, shareType, targetID string) bool {
	var count int
	switch shareType {
	case "collection":
		h.DB.QueryRow(ctx,
			`SELECT COUNT(*) FROM collections WHERE id = $1 AND owner_user_id = $2`,
			targetID, userID,
		).Scan(&count)
	case "file":
		h.DB.QueryRow(ctx,
			`SELECT COUNT(*) FROM files WHERE id = $1 AND uploader_user_id = $2`,
			targetID, userID,
		).Scan(&count)
	}
	return count > 0
}

File: crates/kutup-server/src/handlers/shares.rs
//! Public-share handlers — mirrors `backend/handlers/shares.go`.
//!
//! A public share is a tokenised link to a collection or file. The link key never reaches
//! the server (it lives only in the URL `#fragment`); we store the collection key already
//! wrapped with that link key, so the stored ciphertext is useless without the fragment.
//! Read endpoints are anonymous — the token is the capability.

use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::Json;
use kutup_crypto::drive_envelope::{DriveEnvelopeContextV1, DriveEnvelopePurpose};
use serde::{Deserialize, Serialize};
use serde_json::json;
use time::OffsetDateTime;
use utoipa::ToSchema;
use uuid::Uuid;

use crate::error::{AppError, AppResult};
use crate::handlers::files::{canonical_uuid, validate_envelope};
use crate::handlers::{octet_stream_response, random_token};
use crate::middleware::AuthUser;
use crate::AppState;

#[derive(Debug, Default, Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct CreateShareRequest {
    /// V1 accepts only "collection".
    share_type: String,
    target_id: String,
    collection_key_envelope: String,
    expires_in_hours: Option<i64>,
}

/// `POST /api/share` — mirrors `CreatePublicShare`. The link key is never sent here.
#[utoipa::path(
    post,
    path = "/api/share",
    tag = "shares",
    security(("BearerAuth" = [])),
    request_body = crate::models::CreateShareRequest,
    responses((status = 201, description = "Share created", body = crate::models::CreateShareResult))
)]
pub async fn create_public_share(
    State(state): State<AppState>,
    user: AuthUser,
    Json(req): Json<CreateShareRequest>,
) -> AppResult<Response> {
    let user_id =
        Uuid::parse_str(&user.user_id).map_err(|_| AppError::internal("invalid user id"))?;

    if req.share_type != "collection" {
        return Err(AppError::bad_request("unsupported public share type"));
    }
    let target_uuid = canonical_uuid(&req.target_id)?;
    let key_epoch: Option<i32> = sqlx::query_scalar(
        "SELECT key_epoch FROM collections \
         WHERE id = $1 AND owner_user_id = $2 AND deleted_at IS NULL",
    )
    .bind(target_uuid)
    .bind(user_id)
    .fetch_optional(&state.pool)
    .await?;
    let Some(key_epoch) = key_epoch else {
        return Err(AppError::forbidden("forbidden"));
    };
    let epoch =
        u32::try_from(key_epoch).map_err(|_| AppError::conflict("invalid collection epoch"))?;
    let envelope_context = DriveEnvelopeContextV1::new(
        DriveEnvelopePurpose::PublicLinkCollectionKey,
        epoch,
        1,
        &req.target_id,
        &user_id.to_string(),
    )
    .map_err(|_| AppError::bad_request("invalid Drive envelope"))?;
    validate_envelope(&req.collection_key_envelope, envelope_context)?;

    let token = random_token(32);
    let expires_at: Option<OffsetDateTime> = req
        .expires_in_hours
        .map(|h| OffsetDateTime::now_utc() + time::Duration::hours(h));

    let id: Uuid = sqlx::query_scalar(
        r#"INSERT INTO public_shares (share_type, target_id, token,
                                      collection_key_envelope, collection_key_epoch,
                                      owner_user_id, expires_at)
           VALUES ($1,$2,$3,$4,$5,$6,$7)
           RETURNING id"#,
    )
    .bind(&req.share_type)
    .bind(target_uuid)
    .bind(&token)
    .bind(&req.collection_key_envelope)
    .bind(key_epoch)
    .bind(user_id)
    .bind(expires_at)
    .fetch_one(&state.pool)
    .await
    .map_err(|_| AppError::internal("internal error"))?;

    Ok((StatusCode::CREATED, Json(json!({"id": id, "token": token}))).into_response())
}

/// Authenticated public context plus the opaque key envelope. The fragment key
/// is deliberately absent.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct PublicShareResponse {
    id: Uuid,
    share_type: String,
    target_id: Uuid,
    collection_key_envelope: String,
    collection_key_epoch: i32,
    owner_user_id: Uuid,
    #[serde(with = "time::serde::rfc3339::option")]
    expires_at: Option<OffsetDateTime>,
}

/// `GET /api/share/{token}` — mirrors `GetPublicShare`. Anonymous.
#[utoipa::path(
    get,
    path = "/api/share/{token}",
    tag = "shares",
    params(("token" = String, Path, description = "Share token (the capability)")),
    responses((status = 200, description = "Share metadata + wrapped key", body = crate::models::PublicShareResponse))
)]
pub async fn get_public_share(
    State(state): State<AppState>,
    Path(token): Path<String>,
) -> AppResult<Response> {
    type ShareRow = (
        Uuid,
        String,
        Uuid,
        String,
        i32,
        Uuid,
        Option<OffsetDateTime>,
    );
    let row: Option<ShareRow> = sqlx::query_as(
        r#"SELECT id, share_type, target_id,
                  collection_key_envelope, collection_key_epoch, owner_user_id, expires_at
           FROM public_shares WHERE token = $1"#,
    )
    .bind(&token)
    .fetch_optional(&state.pool)
    .await
    .ok()
    .flatten();
    let Some((id, share_type, target_id, envelope, epoch, owner_user_id, expires_at)) = row else {
        return Err(AppError::not_found("not found"));
    };
    if let Some(exp) = expires_at {
        if OffsetDateTime::now_utc() > exp {
            return Err(AppError::new(StatusCode::GONE, "link expired"));
        }
    }
    Ok(Json(PublicShareResponse {
        id,
        share_type,
        target_id,
        collection_key_envelope: envelope,
        collection_key_epoch: epoch,
        owner_user_id,
        expires_at,
    })
    .into_response())
}

/// One file in a public-collection share. Field order mirrors the Go struct; `created_at`
/// is the Postgres timestamp rendered as the same text Go's `time.Time` JSON produces.
#[derive(Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
struct PublicFileRow {
    id: Uuid,
    collection_id: Uuid,
    metadata_envelope: String,
    file_key_envelope: String,
    key_epoch: i32,
    metadata_revision: i64,
    encrypted_size_bytes: i64,
    #[serde(with = "time::serde::rfc3339")]
    created_at: OffsetDateTime,
}

/// `GET /api/share/{token}/files` — mirrors `ListPublicShareFiles`. Anonymous.
#[utoipa::path(
    get,
    path = "/api/share/{token}/files",
    tag = "shares",
    params(("token" = String, Path, description = "Share token (the capability)")),
    responses((status = 200, description = "Files in the shared collection", body = Vec<PublicFileRow>))
)]
pub async fn list_public_share_files(
    State(state): State<AppState>,
    Path(token): Path<String>,
) -> AppResult<Response> {
    let meta: Option<(Uuid, String, Option<OffsetDateTime>)> = sqlx::query_as(
        "SELECT target_id, share_type, expires_at FROM public_shares WHERE token = $1",
    )
    .bind(&token)
    .fetch_optional(&state.pool)
    .await
    .ok()
    .flatten();
    let Some((target_id, share_type, expires_at)) = meta else {
        return Err(AppError::not_found("not found"));
    };
    if let Some(exp) = expires_at {
        if OffsetDateTime::now_utc() > exp {
            return Err(AppError::new(StatusCode::GONE, "link expired"));
        }
    }
    if share_type != "collection" {
        return Err(AppError::bad_request("not a collection share"));
    }
    // A trashed collection's share links go dark until it is restored.
    let live: Option<i64> =
        sqlx::query_scalar("SELECT COUNT(*) FROM collections WHERE id = $1 AND deleted_at IS NULL")
            .bind(target_id)
            .fetch_optional(&state.pool)
            .await
            .ok()
            .flatten();
    if live.unwrap_or(0) == 0 {
        return Err(AppError::not_found("not found"));
    }

    type PubFileTuple = (Uuid, Uuid, String, String, i32, i64, i64, OffsetDateTime);
    let rows: Vec<PubFileTuple> = sqlx::query_as(
        r#"SELECT id, collection_id, metadata_envelope, file_key_envelope,
                  key_epoch, metadata_revision, encrypted_size_bytes, created_at
           FROM files WHERE collection_id = $1 AND deleted_at IS NULL
           ORDER BY created_at DESC"#,
    )
    .bind(target_id)
    .fetch_all(&state.pool)
    .await
    .map_err(|_| AppError::internal("internal error"))?;

    let files: Vec<PublicFileRow> = rows
        .into_iter()
        .map(
            |(id, collection_id, metadata, file_key, epoch, revision, size, created_at)| {
                PublicFileRow {
                    id,
                    collection_id,
                    metadata_envelope: metadata,
                    file_key_envelope: file_key,
                    key_epoch: epoch,
                    metadata_revision: revision,
                    encrypted_size_bytes: size,
                    created_at,
                }
            },
        )
        .collect();
    Ok(Json(files).into_response())
}

/// `GET /api/share/{token}/download/{fileId}` — streams the encrypted blob. Anonymous:
/// the token is the capability, and the content is E2EE (the link key that unwraps it
/// lives only in the URL fragment, never reaching the server).
///
/// This used to return a presigned S3 URL, but the storage endpoint is deliberately
/// unreachable from outside the deployment (`http://seaweedfs-s3:8333` in the bundled
/// compose), so no external client could follow it. Streaming through the backend
/// matches every other download path.
#[utoipa::path(
    get,
    path = "/api/share/{token}/download/{fileId}",
    tag = "shares",
    params(
        ("token" = String, Path, description = "Share token (the capability)"),
        ("fileId" = String, Path, description = "File id")
    ),
    responses((status = 200, description = "The encrypted blob (application/octet-stream)"))
)]
pub async fn download_public_share_file(
    State(state): State<AppState>,
    Path((token, file_id)): Path<(String, String)>,
) -> AppResult<Response> {
    let meta: Option<(Uuid, String, Option<OffsetDateTime>)> = sqlx::query_as(
        "SELECT target_id, share_type, expires_at FROM public_shares WHERE token = $1",
    )
    .bind(&token)
    .fetch_optional(&state.pool)
    .await
    .ok()
    .flatten();
    let Some((target_id, share_type, expires_at)) = meta else {
        return Err(AppError::not_found("not found"));
    };
    if let Some(exp) = expires_at {
        if OffsetDateTime::now_utc() > exp {
            return Err(AppError::new(StatusCode::GONE, "link expired"));
        }
    }

    let fid = Uuid::parse_str(&file_id).map_err(|_| AppError::not_found("not found"))?;
    let file: Option<(String, Uuid)> = sqlx::query_as(
        "SELECT storage_path, collection_id FROM files WHERE id = $1 AND deleted_at IS NULL",
    )
    .bind(fid)
    .fetch_optional(&state.pool)
    .await
    .ok()
    .flatten();
    let Some((storage_path, coll_id)) = file else {
        return Err(AppError::not_found("not found"));
    };

    if share_type == "collection" && coll_id != target_id {
        return Err(AppError::forbidden("forbidden"));
    }
    if share_type == "file" && fid != target_id {
        return Err(AppError::forbidden("forbidden"));
    }

    let (body, size) = state
        .storage
        .get_object(&storage_path)
        .await
        .map_err(|_| AppError::internal("storage"))?;
    Ok(octet_stream_response(body, size, &[]))
}
